### PR TITLE
fix(evalhub): detect pods stuck in Pending due to missing PVC and mark job as FAILED

### PR DESCRIPTION
**Pull request by scheruku-rh**
When an eval job references a PVC that doesn't exist, the Kubernetes scheduler sets `PodScheduled=False/Unschedulable` on the pod but never starts it. Because no container runs, the adapter cannot POST a failure event back to EvalHub, the job stays in a running state with no user-visible error.
The existing `jobStatusIndicatesTerminalFailure()` check requires Active=0, Failed>0 on the Job. A pod stuck in Pending keeps Active=1, Failed=0, so that path never fires. This PR extends the pod predicate path instead.

Testing:
1. created an evaluation job with non-existing pvc:

```
{
  "resource": {
    "id": "d33ae87a-a8ae-47d4-8fd7-0acf2323cc2d",
    "tenant": "tenant",
    "created_at": "2026-07-13T10:50:49.64252227Z",
    "owner": "system:serviceaccount:tenant:tenant-user",
    "mlflow_experiment_id": "1"
  },
  "status": {
    "state": "pending",
    "message": {
      "message": "Evaluation job created",
      "message_code": "evaluation_job_created"
    }
  },
  "results": {
    "mlflow_experiment_url": "https://mlflow.opendatahub.svc.cluster.local:8443/api/2.0/mlflow/experiments"
  },
  "name": "model api key/cert test",
  "model": {
    "url": "https://vllm-llama-prabhu.apps.rosa.prabhu-comhub.xqmp.p3.openshiftapps.com/v1",
    "name": "meta-llama/Llama-3.2-1B-Instruct",
    "auth": {
      "secret_ref": "vllm-api-key"
    }
  },
  "benchmarks": [
    {
      "id": "arc_easy",
      "provider_id": "lm_evaluation_harness",
      "parameters": {
        "limit": 5
      },
      "test_data_ref": {
        "pvc": {
          "claim_name": "my-test-storage-pvc-in-prabhu-ns"
        }
      }
    }
  ],
  "experiment": {
    "name": "exp12"
  }
}
```

2. Logs show that job is marked as FAILED after about 2min as expected

```
2026-07-13T10:50:49Z	INFO	pod unschedulable within grace period — requeuing	{"controller": "evalhub-evaluation-job-failure", "controllerGroup": "batch", "controllerKind": "Job", "Job": {"name":"d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98","namespace":"tenant"}, "namespace": "tenant", "name": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "reconcileID": "3d041298-2edf-4487-9e51-25931c9a7ff7", "evalhub_watcher": "failure_sync", "controller": "evalhub-evaluation-job-failure", "action": "requeue_scheduling_grace", "job": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "namespace": "tenant", "requeueAfter": "1m59.243314458s"}
2026-07-13T10:52:49Z	INFO	reconcile start	{"controller": "evalhub-evaluation-job-failure", "controllerGroup": "batch", "controllerKind": "Job", "Job": {"name":"d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98","namespace":"tenant"}, "namespace": "tenant", "name": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "reconcileID": "217977fb-1fdf-42dd-a8ac-b6424a527353", "evalhub_watcher": "failure_sync", "controller": "evalhub-evaluation-job-failure", "action": "reconcile", "namespace": "tenant", "name": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98"}

2026-07-13T10:52:49Z	INFO	operator-only failure detected	{"controller": "evalhub-evaluation-job-failure", "controllerGroup": "batch", "controllerKind": "Job", "Job": {"name":"d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98","namespace":"tenant"}, "namespace": "tenant", "name": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "reconcileID": "217977fb-1fdf-42dd-a8ac-b6424a527353", "evalhub_watcher": "failure_sync", "controller": "evalhub-evaluation-job-failure", "action": "failure_detected", "job": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "namespace": "tenant", "detail": "pod d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5lbc9n: pod unschedulable: 0/1 nodes are available: persistentvolumeclaim \"my-test-storage-pvc-in-prabhu-ns\" not found. preemption: 0/1 nodes are available: 1 Preemption is not helpful for scheduling."}
2026-07-13T10:52:49Z	INFO	deleted evaluation job after EvalHub failure sync	{"controller": "evalhub-evaluation-job-failure", "controllerGroup": "batch", "controllerKind": "Job", "Job": {"name":"d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98","namespace":"tenant"}, "namespace": "tenant", "name": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "reconcileID": "217977fb-1fdf-42dd-a8ac-b6424a527353", "evalhub_watcher": "failure_sync", "controller": "evalhub-evaluation-job-failure", "action": "delete_job_ok", "job": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "namespace": "tenant"}
2026-07-13T10:52:49Z	INFO	posted EvalHub benchmark failure event	{"controller": "evalhub-evaluation-job-failure", "controllerGroup": "batch", "controllerKind": "Job", "Job": {"name":"d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98","namespace":"tenant"}, "namespace": "tenant", "name": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "reconcileID": "217977fb-1fdf-42dd-a8ac-b6424a527353", "evalhub_watcher": "failure_sync", "controller": "evalhub-evaluation-job-failure", "action": "post_events_ok", "evalJobID": "d33ae87a-a8ae-47d4-8fd7-0acf2323cc2d", "k8sJob": "d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5e4d98", "namespace": "tenant"}

```

3. Job is updated to FAILED:
```
{
  "resource": {
    "id": "d33ae87a-a8ae-47d4-8fd7-0acf2323cc2d",
    "tenant": "tenant",
    "created_at": "2026-07-13T10:50:49.642618Z",
    "updated_at": "2026-07-13T10:52:49.048111Z",
    "owner": "system:serviceaccount:tenant:tenant-user",
    "mlflow_experiment_id": "1"
  },
  "status": {
    "state": "failed",
    "message": {
      "message": "Evaluation job is failed. \nBenchmark arc_easy failed with message: pod d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5lbc9n: pod unschedulable: 0/1 nodes are available: persistentvolumeclaim \"my-test-storage-pvc-in-prabhu-ns\" not found. preemption: 0/1 nodes are available: 1 Preemption is not helpful for scheduling.\n",
      "message_code": "evaluation_job_updated"
    },
    "benchmarks": [
      {
        "provider_id": "lm_evaluation_harness",
        "id": "arc_easy",
        "benchmark_index": 0,
        "status": "failed",
        "error_message": {
          "message": "pod d33ae87a-a8ae-47d4-8fd7-0a-56b6bf13-7973-460e-8fae-4a16cd5lbc9n: pod unschedulable: 0/1 nodes are available: persistentvolumeclaim \"my-test-storage-pvc-in-prabhu-ns\" not found. preemption: 0/1 nodes are available: 1 Preemption is not helpful for scheduling.",
          "message_code": "RUNTIME_FAILURE"
        }
      }
    ]
  },
  "results": {
    "benchmarks": [
      {
        "id": "arc_easy",
        "provider_id": "lm_evaluation_harness",
        "benchmark_index": 0
      }
    ],
    "mlflow_experiment_url": "https://mlflow.opendatahub.svc.cluster.local:8443/api/2.0/mlflow/experiments"
  },
  "name": "model api key/cert test",
  "model": {
    "url": "https://vllm-llama-prabhu.apps.rosa.prabhu-comhub.xqmp.p3.openshiftapps.com/v1",
    "name": "meta-llama/Llama-3.2-1B-Instruct",
    "auth": {
      "secret_ref": "vllm-api-key"
    }
  },
  "benchmarks": [
    {
      "id": "arc_easy",
      "provider_id": "lm_evaluation_harness",
      "parameters": {
        "limit": 5
      },
      "test_data_ref": {
        "pvc": {
          "claim_name": "my-test-storage-pvc-in-prabhu-ns"
        }
      }
    }
  ],
  "experiment": {
    "name": "exp12"
  }
}

```

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling for pods stuck `Pending` with `PodScheduled=False`/`Unschedulable` by introducing a configurable scheduling grace period and adjusting requeue decisions until the grace window expires.
  * Refined operator-only failure detection and messaging for unschedulable pending pods, including PVC context when available and a fallback message when details are missing.

* **Tests**
  * Expanded unit tests to validate grace-period timing behavior, operator-only detection rules, and scheduling helper edge cases (including missing or zero transition times).

### CODE REVIEW

**Review by gnaulak-redhat — APPROVED** [review]
lgtm afaict

File: controllers/evalhub/evaluation_job_failure_reconciler.go [hunk 1]
```diff
@@ -498,9 +498,33 @@ func podIndicatesOperatorOnlyFailure(pod *corev1.Pod) bool {
 	return ok
 }
 
+// podSchedulingFailureMessage returns a non-empty message and true when the pod is stuck in Pending
+// because the scheduler could not place it (e.g. a referenced PVC does not exist).
+// The pod never starts, so no container callback to EvalHub is possible.
+func podSchedulingFailureMessage(pod *corev1.Pod) (string, bool) {
```
Comment (ruivieira) [inline controllers/evalhub/evaluation_job_failure_reconciler.go:534]: @scheruku-rh  IIUC this fires on the very first reconcile, and `PodReasonUnschedulable` is used for both impossible conditions (missing PVC) and transient ones (insufficient resources while a node scales up), so we could get false positives on auto-scaling clusters. What do you think about adding a grace period using the condition's `LastTransitionTime`? Something like:

```go
  const schedulingGracePeriod = 2 * time.Minute

  func podSchedulingFailureMessage(pod *corev1.Pod) (string, bool) {
        if pod.Status.Phase != corev1.PodPending {
                return "", false
        }
        for _, c := range pod.Status.Conditions {
                if c.Type == corev1.PodScheduled &&
                        c.Status == corev1.ConditionFalse &&
                        c.Reason == corev1.PodReasonUnschedulable &&
                        time.Since(c.LastTransitionTime.Time) > schedulingGracePeriod {
                        msg := c.Message
                        if msg == "" {
                                msg = "pod unschedulable"
                        }
                        return fmt.Sprintf("pod unschedulable: %s", msg), true
                }
        }
        return "", false
  }
```

That way a pod that's transiently unschedulable during node scale-up gets a couple of minutes to resolve before we mark it as failed. The reconciler already requeues so no extra plumbing needed. The check just won't match until the grace period elapses. Could also be a follow-up if you'd rather do it on separate PR.

wdyt?

**Review by ruivieira — CHANGES_REQUESTED** [review]

**Review by ruivieira — APPROVED** [review]
